### ISSUE TITLE
chap04_Classification_softmax_classifier_iris

### ISSUE THREAD

**Issue by hih0124** [hih0124/lecture_deep_learning#34]
"""
softmax + iris
"""
import tensorflow.compat.v1 as tf # ver 1.x
tf.disable_v2_behavior() # ver 2.x 사용 안 함
from sklearn.datasets import load_iris
from sklearn.preprocessing import OneHotEncoder # y data 원핫인코딩하기 위해서
from sklearn.metrics import accuracy_score
import numpy as np

### 1. x, y 공급 data 
iris = load_iris()

-x변수 : 1~4 칼럼
x_data = iris.data
x_data.shape # (150, 4)

-y변수 
y_data = iris.target
y_data.shape # (150, 1)

- reshape
y_data = y_data.reshape(-1, 1)
Y = tf.placeholder(dtype=tf.float32, shape = [None, 1])에 적재할수 있도록 reshape
y_data.shape # (150, 1)

'''
0 -> 1 0 0
1 -> 0 1 0
2 -> 0 0 1
'''

obj = OneHotEncoder()
-y_data = obj.fit_transform(y_data)
-y_data
'''
<150x3 sparse matrix of type '<class 'numpy.float64'>'
	with 150 stored elements in Compressed Sparse Row format>
'''
-sparse -> numpy
y_data = obj.fit_transform(y_data).toarray()
y_data.shape # (150, 3)

### 2. X,Y 변수 정의
X = tf.placeholder(dtype=tf.float32, shape = [None, 4]) # [관측치, 입력수]
Y = tf.placeholder(dtype=tf.float32, shape = [None, 3]) # [관측치, 출력수]

### 3. w,b 변수 정의
w = tf.Variable(tf.random_normal([4, 3])) # [입력수, 출력수]
b = tf.Variable(tf.random_normal([3])) # [출력수]

### 4.softmax 분류기 
- 1) 회귀방정식 : 예측치 
model = tf.matmul(X, w) + b # 회귀모델 

- softmax(예측치)
softmax = tf.nn.softmax(model) # 활성함수 적용 y1:0.8, y2:0.1, y3:0.1

- (2) loss function : Cross Entropy 이용 : -sum(Y * log(model)) 
loss = -tf.reduce_mean(Y * tf.log(softmax) + (1 - Y) * tf.log(1 - softmax))

- 2차 방법 : softmax + crossEntropy
model 다음에 바로 이거 쓰면 된다. ## 얘는 - 부호 없음
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(
    labels = Y, logits = model))

- 3) optimizer : 오차 최소화(w, b update) 
train = tf.train.AdamOptimizer(0.1).minimize(loss) # 오차 최소화

- 4) argmax() : encoding(2진수) -> decoding(10진수로)
y_pred = tf.argmax(softmax, axis = 1)
y_true = tf.argmax(Y, axis = 1)

- 5. model 학습
with tf.Session() as sess : 
    sess.run(tf.global_variables_initializer()) # w,b 초기화
    
    feed_data = {X : x_data, Y : y_data}
    
    - 반복학습 : 500회
    for step in range(500) :
        _, loss_val = sess.run([train, loss], feed_dict = feed_data)
        
        if (step+1) % 50 == 0 :
            print("step={}, loss ={}".format(step+1, loss_val))
    
    - model result
    print(sess.run(softmax, feed_dict = {X : x_data}))
    y_pred_re = sess.run(y_pred, feed_dict = {X : x_data}) # 예측치
    y_true_re = sess.run(y_true, feed_dict = {Y : y_data}) # 정답
    
    print("y pred=", y_pred_re)
    print("y true = ", y_true_re)
    acc= accuracy_score(y_true_re, y_pred_re)
    print("accuracy =", acc) # accuracy = 0.98
    
    y_pred_re[:20]
    '''
    array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
      dtype=int64)
    '''
    y_true_re[:20]
    '''
    array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
      dtype=int64)
    '''
    
    import matplotlib.pyplot as plt
    plt.plot(y_pred_re, color='r')
    plt.plot(y_true_re, color='b')
    plt.show()
